notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/TemplateVisitor.java

```diff
@@ -0,0 +1,217 @@
+        return true;
+    }
+
+    /**
+     * Deletes one template from this collection by name.
+     *
+     * @param templateName The name of template that should no longer be in the collection.
+     * @return True if the template had previously been associated with this collection.
+     */
+    public boolean removeTemplate(String templateName) {
+        return removeTemplates(Collections.singletonList(templateName));
+    }
+
+    /**
+     * Deletes several templates from this collection.
+     *
+     * @param templates The templates name collection that should no longer be in the collection.
+     * @return True if any of the templates had previously been associated with this collection.
+     */
+    public boolean removeTemplates(List<String> templates) {
+        Map<String, InstallationTemplate> sharedPreferenceTemplates = getSharedPreferenceTemplates();
+        sharedPreferenceTemplates.keySet().removeAll(templates);
+        Set<String> serializedTemplatesSet = new HashSet<>();
+        for (Map.Entry<String, InstallationTemplate> installation: sharedPreferenceTemplates.entrySet()) {
+            serializedTemplatesSet.add(serializeInstallationTemplateToJson(installation.getKey(), installation.getValue()));
+        }
+        mPreferences.edit().putStringSet(PREFERENCE_KEY, serializedTemplatesSet).apply();
+        return true;
+    }
+
+    /**
+     * Fetches the template by name
+     * @param templateName Name of template
+     * @return Return template associated with name
+     */
+    public InstallationTemplate getTemplate(String templateName) {
+        Map<String, InstallationTemplate> templates = getSharedPreferenceTemplates();
+        return templates.get(templateName);
+    }
+
+    /**
+     * Fetches the templates associated with this collection.
+     *
+     * @return A set of templates.
+     */
+    public Iterable<Map.Entry<String, InstallationTemplate>> getTemplates() {
+        return getSharedPreferenceTemplates().entrySet();
+    }
+
+    /**
+     * Empties the collection of templates.
+     */
+    public void clearTemplates() {
+        mPreferences.edit().remove(PREFERENCE_KEY).apply();
+    }
+
+    /**
+     * Serialize InstallationTemplate to JSONObject.
+     *
+     * @param name Name of template
+     * @param installationTemplate The templates that should no longer be in the collection.
+     * @return serialized templateObject.
+     */
+    private static String serializeInstallationTemplateToJson(String name, InstallationTemplate installationTemplate) {
+        JSONObject templateObject = new JSONObject();
+        try {
+            templateObject.put("name", name);
+            templateObject.put("body", installationTemplate.getBody());
+            JSONObject headers = new JSONObject();
+            Iterator<Map.Entry<String, String>> headersIterators = installationTemplate.getHeaders().iterator();
+            while (headersIterators.hasNext()) {
```

Comment:
Total stylistic nit. Why explicitly new-up an iterator instead of using a `foreach` loop?